app/upgrades/v20/upgrades_test.go

```diff
@@ -0,0 +1,290 @@
+package v20_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/v19/app/apptesting"
+	v20 "github.com/osmosis-labs/osmosis/v19/app/upgrades/v20"
+	gammmigration "github.com/osmosis-labs/osmosis/v19/x/gamm/types/migration"
+	poolincentivestypes "github.com/osmosis-labs/osmosis/v19/x/pool-incentives/types"
+)
+
+type UpgradeTestSuite struct {
+	apptesting.KeeperTestHelper
+}
+
+func (s *UpgradeTestSuite) SetupTest() {
+	s.Setup()
+}
+
+func TestUpgradeTestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeTestSuite))
+}
+
+// Validates that the pool incentive distribution records are updated correctly.
+// The expected update is to convert concentrated gauges to group gauges iff
+// - migration record exists for the concentrated pool and another CFMM pool
+// - if migration between concentrated and CFMM exists, then the CFMM pool is not incentivized individually
+// It is expected that all other distribution records are not modified.
+func (s *UpgradeTestSuite) TestCreateGroupsForIncentivePairs() {
```

Comment:
I might of missed it but I think we still need a "migration link exists, but for some reason we we still have the balancer pool record as well as the CL pool record in the distrRecords". This is technically valid state that is possible to exist on mainnet.

Reply:
This is covered by this test:
https://github.com/osmosis-labs/osmosis/blob/6147a72361e5b9ce954d0c064e3fe83962b7ed89/app/upgrades/v20/upgrades_test.go#L191-L220

Please let me know if there is another test you were thinking about

Reply:
Added: [0c159ce](https://github.com/osmosis-labs/osmosis/pull/6610/commits/0c159ceb73845e7a0b90aa2df1c3b5b90c488c1a)